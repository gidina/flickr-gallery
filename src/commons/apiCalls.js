import { API_KEY, GALLERY_ID } from "../config";

const API_URL_GET_PHOTOS = `https://api.flickr.com/services/rest/?method=flickr.galleries.getPhotos&api_key=${API_KEY}&gallery_id=${GALLERY_ID}&extras=description%2C+media%2C+owner_name&format=json&nojsoncallback=1`;
const API_URL_GET_PHOTOS_SIZES = photoId =>
  `https://api.flickr.com/services/rest/?method=flickr.photos.getSizes&api_key=${API_KEY}&photo_id=${photoId}&format=json&nojsoncallback=1`;

export const getPhotos = () => 
  fetch(API_URL_GET_PHOTOS)
    .then(response => response.json())
    .then(json => json.photos.photo);

export const getPhotosSizes = (photoId) => 
  fetch(API_URL_GET_PHOTOS_SIZES(photoId))
  .then(response => response.json())
  .then(json => json.sizes.size);

