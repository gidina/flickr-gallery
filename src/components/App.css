/*  SCROLLBAR
-------------------*/
::-webkit-scrollbar {
  width: 0.2em;
  background: #e74c3c;
}

::-webkit-scrollbar-thumb {
  background: #fff;
}

::-webkit-scrollbar-thumb:window-inactive {
  background: #fff;
}

body {
  background: #EEE;
  overflow-X: hidden;
  font-size: 13px;
  font-family: 'Metrophobic', sans-serif;
  font-weight: 400;
}
