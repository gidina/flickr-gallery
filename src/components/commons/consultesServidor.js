export const getPhotos = (url) => 
    fetch(url)
      .then(response => response.json())
      .then(json => json);

export const getPhotosSizes = (url) => 
  fetch(API_URL_GET_PHOTOS_SIZES(photo.id))
  .then(response => response.json())
  .then(json => json);

