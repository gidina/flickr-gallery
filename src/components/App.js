import React, { Component, Fragment } from "react";
import Gallery from "./Gallery";
import Loader from "./Loader";
import Modal from "./Modal";
import { API_KEY, redColor } from "../config";
import "./App.css";

const GALLERY_ID = "117615905-72157695735361740";
const API_URL_GET_PHOTOS = `https://api.flickr.com/services/rest/?method=flickr.galleries.getPhotos&api_key=${API_KEY}&gallery_id=${GALLERY_ID}&extras=description%2C+media%2C+owner_name&format=json&nojsoncallback=1`;
const API_URL_GET_PHOTOS_SIZES = photoId =>
  `https://api.flickr.com/services/rest/?method=flickr.photos.getSizes&api_key=${API_KEY}&photo_id=${photoId}&format=json&nojsoncallback=1`;

const getSourceByParam = (array, parametreCerca) =>
  array.find(size => size.label === parametreCerca).source;

class App extends Component {
  state = {
    isLoading: true,
    fotosGaleria: null,
    fotoSeleccionada: null,
    isLoadingModal: false
  };
  componentDidMount = () => {
    fetch(API_URL_GET_PHOTOS)
      .then(response => response.json())
      .then(json =>
        this.setState({
          ...this.state,
          isLoading: false,
          fotosGaleria: json.photos.photo
        })
      );
  };
  onClickCloseModal = () => {
    this.setState({ ...this.state, fotoSeleccionada: null });
  };
  onClickPhotoItemHandler = photo => {
    const parametreCerca = photo.media === "video" ? "Video Player" : "Large";

    const redirectURL = `https://www.flickr.com/photos/${photo.owner}/${photo.id}/in/gallery-${GALLERY_ID}/`;
    this.setState({
      ...this.state,
      isLoadingModal: true,
      fotoSeleccionada: {
        title: photo.title,
        description: photo.description._content,
        redirectURL,
        isVideo: photo.media === "video"
      }
    });

    fetch(API_URL_GET_PHOTOS_SIZES(photo.id))
      .then(response => response.json())
      .then(json => {
        this.setState({
          ...this.state,
          isLoadingModal: false,
          fotoSeleccionada: {
            ...this.state.fotoSeleccionada,
            source: getSourceByParam(json.sizes.size, parametreCerca)
          }
        });
      });
  };
  render = () => {
    const {
      isLoading,
      fotosGaleria,
      fotoSeleccionada,
      isLoadingModal
    } = this.state;
    return (
      <Fragment>
        {isLoading ? (
          <Loader color={redColor} />
        ) : (
          <Fragment>
            <Gallery
              photos={fotosGaleria}
              onClickPhoto={this.onClickPhotoItemHandler}
            />
            {fotoSeleccionada ? (
              <Modal
                isLoading={isLoadingModal}
                photo={fotoSeleccionada}
                onClose={this.onClickCloseModal}
              />
            ) : null}
          </Fragment>
        )}
      </Fragment>
    );
  };
}

export default App;
