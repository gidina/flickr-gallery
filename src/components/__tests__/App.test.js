import React from "react";
import renderer from "react-test-renderer";
import App from "../App";
import { getPhotos, getPhotosSizes } from "../commons/consultesServidor";

jest.mock("../commons/consultesServidor");

jest.mock("../Loader");
jest.mock("../Modal");
jest.mock("../Gallery");

const photosTest = [
  {
    id: 1,
    title: "Photo 1",
    description: { _content: "Description Photo 1" },
    media: "video",
    farm: 1,
    server: "server1",
    secret: 123456
  },
  {
    id: 2,
    title: "Photo 2",
    description: { _content: "Description Photo 2" },
    media: "video",
    farm: 1,
    server: "server1",
    secret: 456789
  }
];

const sizesTest = [
  {
    label: "Video Player",
    source: "source Video Player size"
  },
  {
    label: "Large",
    source: "source Large size"
  },
];

beforeEach(() => {
  jest.clearAllMocks();
});

test("Si isLoading és true, es mostra un Loader de color red (#f55)", () => {
  getPhotos.mockImplementation(() => Promise.resolve(photosTest));

  expect(getPhotos).toHaveBeenCalledTimes(0);
  const tree = renderer
    .create(<App />)
    .toJSON();
  expect(tree).toMatchSnapshot();
  expect(getPhotos).toHaveBeenCalledTimes(1);
  expect(getPhotos).toHaveBeenCalledWith();
});

test("Si no s'ha seleccionat cap foto, no es mostra el Modal", async () => {
  getPhotos.mockImplementation(() => Promise.resolve(photosTest));

  const tree = await renderer.create(<App />);

  expect(tree.toJSON()).toMatchSnapshot();
});

test("Si s'ha seleccionat una foto, es mostra el Modal amb la propietat isLoading a true", async () => {
  const photoTests = photosTest[0];
  getPhotos.mockImplementation(() => Promise.resolve(photosTest));
  getPhotosSizes.mockImplementation(() => Promise.resolve(sizesTest));

  let tree = await renderer.create(<App />);

  expect(getPhotosSizes).toHaveBeenCalledTimes(0);

  tree.getInstance().onClickPhotoItemHandler(photoTests);

  expect(tree.toJSON()).toMatchSnapshot();
  expect(getPhotosSizes).toHaveBeenCalledTimes(1);
  expect(getPhotosSizes).toHaveBeenCalledWith(photoTests.id);
});

test("Si s'ha seleccionat una foto i s'ha acabat la petició per obtenir les mides, es mostra el Modal", async () => {
  const photoTests = photosTest[0];
  getPhotos.mockImplementation(() => Promise.resolve(photosTest));
  getPhotosSizes.mockImplementation(() => Promise.resolve(sizesTest));

  let tree = renderer.create(<App />);

  await tree.getInstance().onClickPhotoItemHandler(photoTests);

  expect(tree.toJSON()).toMatchSnapshot();
});

test("Si es rep l'onClose del modal, es deselecciona la foto", async () => {
  const photoTests = photosTest[0];
  getPhotos.mockImplementation(() => Promise.resolve(photosTest));
  getPhotosSizes.mockImplementation(() => Promise.resolve(sizesTest));

  let tree = renderer.create(<App />);

  await tree.getInstance().onClickPhotoItemHandler(photoTests);
  tree.getInstance().onClickCloseModal();

  expect(tree.toJSON()).toMatchSnapshot();
});