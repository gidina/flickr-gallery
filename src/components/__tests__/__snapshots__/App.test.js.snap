// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Si es rep l'onClose del modal, es deselecciona la foto 1`] = `
<gallery-mock
  onClickPhoto={[Function]}
  photos={
    Array [
      Object {
        "description": Object {
          "_content": "Description Photo 1",
        },
        "farm": 1,
        "id": 1,
        "media": "video",
        "secret": 123456,
        "server": "server1",
        "title": "Photo 1",
      },
      Object {
        "description": Object {
          "_content": "Description Photo 2",
        },
        "farm": 1,
        "id": 2,
        "media": "video",
        "secret": 456789,
        "server": "server1",
        "title": "Photo 2",
      },
    ]
  }
/>
`;

exports[`Si isLoading és true, es mostra un Loader de color red (#f55) 1`] = `
<loader-mock
  color="#f55"
/>
`;

exports[`Si no s'ha seleccionat cap foto, no es mostra el Modal 1`] = `
<gallery-mock
  onClickPhoto={[Function]}
  photos={
    Array [
      Object {
        "description": Object {
          "_content": "Description Photo 1",
        },
        "farm": 1,
        "id": 1,
        "media": "video",
        "secret": 123456,
        "server": "server1",
        "title": "Photo 1",
      },
      Object {
        "description": Object {
          "_content": "Description Photo 2",
        },
        "farm": 1,
        "id": 2,
        "media": "video",
        "secret": 456789,
        "server": "server1",
        "title": "Photo 2",
      },
    ]
  }
/>
`;

exports[`Si s'ha seleccionat una foto i s'ha acabat la petició per obtenir les mides, es mostra el Modal 1`] = `
Array [
  <gallery-mock
    onClickPhoto={[Function]}
    photos={
      Array [
        Object {
          "description": Object {
            "_content": "Description Photo 1",
          },
          "farm": 1,
          "id": 1,
          "media": "video",
          "secret": 123456,
          "server": "server1",
          "title": "Photo 1",
        },
        Object {
          "description": Object {
            "_content": "Description Photo 2",
          },
          "farm": 1,
          "id": 2,
          "media": "video",
          "secret": 456789,
          "server": "server1",
          "title": "Photo 2",
        },
      ]
    }
  />,
  <modal-mock
    isLoading={false}
    onClose={[Function]}
    photo={
      Object {
        "description": "Description Photo 1",
        "isVideo": true,
        "redirectURL": "https://www.flickr.com/photos/undefined/1/in/gallery-117615905-72157695735361740/",
        "source": "source Video Player size",
        "title": "Photo 1",
        "username": undefined,
      }
    }
  />,
]
`;

exports[`Si s'ha seleccionat una foto, es mostra el Modal amb la propietat isLoading a true 1`] = `
Array [
  <gallery-mock
    onClickPhoto={[Function]}
    photos={
      Array [
        Object {
          "description": Object {
            "_content": "Description Photo 1",
          },
          "farm": 1,
          "id": 1,
          "media": "video",
          "secret": 123456,
          "server": "server1",
          "title": "Photo 1",
        },
        Object {
          "description": Object {
            "_content": "Description Photo 2",
          },
          "farm": 1,
          "id": 2,
          "media": "video",
          "secret": 456789,
          "server": "server1",
          "title": "Photo 2",
        },
      ]
    }
  />,
  <modal-mock
    isLoading={true}
    onClose={[Function]}
    photo={
      Object {
        "description": "Description Photo 1",
        "isVideo": true,
        "redirectURL": "https://www.flickr.com/photos/undefined/1/in/gallery-117615905-72157695735361740/",
        "title": "Photo 1",
        "username": undefined,
      }
    }
  />,
]
`;
